keras/src/callbacks/terminate_on_nan.py

```diff
@@ -7,14 +7,63 @@
 
 @keras_export("keras.callbacks.TerminateOnNaN")
 class TerminateOnNaN(Callback):
-    """Callback that terminates training when a NaN loss is encountered."""
+    """Callback that terminates training when a NaN loss is encountered.
+
+    This callback monitors the loss value during training
+    and terminates training when a NaN or Inf loss is detected.
+    By default, training is stopped gracefully
+    by setting `model.stop_training = True`, which triggers all callback cleanup
+    methods including `on_train_end()`.
+
+    Alternatively, you can use `hard=True` to immediately raise a RuntimeError
+    when NaN/Inf is detected. This hard termination prevents `on_train_end()`
+    from being called on other callbacks, which is useful for preserving backup
+    states or preventing unintended cleanup when training fails.
+
+    Args:
+        hard: Boolean, default False. If False, uses graceful stop via
+            `model.stop_training = True`. If True, immediately raises
+            RuntimeError on NaN/Inf loss, bypassing callback cleanup methods.
+
+    Example:
+
+    ```
+    # Graceful termination (default)
+    callback = keras.callbacks.TerminateOnNaN()
+    model.fit(x, y, callbacks=[callback])
+
+    # Hard termination (strict failure)
+    callback = keras.callbacks.TerminateOnNaN(hard=True)
+    model.fit(x, y, callbacks=[callback])
+    ```
+    """
+
+    def __init__(self, hard: bool = False):
```

Comment:
How about we call the new argument `raise_error` instead of `hard`? Hopefully that name makes it obvious how it's different.